### ISSUE TITLE
main.CRITICAL: Exception: Notice: Undefined index: state

### ISSUE THREAD

**Issue by lnmn249** [amzn/amazon-payments-magento-2-plugin#138]
Running Magento CE 2.1.9 on CentOS 7 with Amazon Payments Plugin
Installed the plugin via the Web Setup Wizard (the preferred documented method)

After attempting to place an order in the Amazon sandbox environment I receive the following error:

An error occurred on the server. Please try to place the order again.

Shortly after I receive an email stated my order is complete, however the order is not in the orders section of the backend.

Screenshot of Chrome JS console. and an excerpt from the exception.log file
Cache was flushed and indexes are set to update via schedule

Regards, 

Dale

* Magento version: 2.1.9
* Magento Edition: Community
* JavaScript Console: Please add a screenshot of the Chrome JavaScript console to speed up the process
![screen shot 2017-10-01 at 4 53 23 pm](https://user-images.githubusercontent.com/6722268/31059519-1ee2be46-a6c9-11e7-9e71-43630b929328.png)

```
[2017-10-01 21:38:35] main.CRITICAL: Exception: Notice: Undefined index: state in /var/www/html/magneto/vendor/amzn/amazon-pay-module/Observer/CompleteOrder.php on line 58 in /var/www/html/magneto/vendor/magento/framework/App/ErrorHandler.php:61
Stack trace:
#0 /var/www/html/magneto/vendor/amzn/amazon-pay-module/Observer/CompleteOrder.php(58): Magento\Framework\App\ErrorHandler->handler(8, 'Undefined index...', '/var/www/html/m...', 58, Array)
#1 /var/www/html/magneto/vendor/magento/framework/Event/Invoker/InvokerDefault.php(73): Amazon\Payment\Observer\CompleteOrder->execute(Object(Magento\Framework\Event\Observer))
#2 /var/www/html/magneto/vendor/magento/framework/Event/Invoker/InvokerDefault.php(61): Magento\Framework\Event\Invoker\InvokerDefault->_callObserverMethod(Object(Amazon\Payment\Observer\CompleteOrder), Object(Magento\Framework\Event\Observer))
#3 /var/www/html/magneto/vendor/magento/framework/Event/Manager.php(66): Magento\Framework\Event\Invoker\InvokerDefault->dispatch(Array, Object(Magento\Framework\Event\Observer))
#4 /var/www/html/magneto/var/generation/Magento/Framework/Event/Manager/Proxy.php(95): Magento\Framework\Event\Manager->dispatch('sales_order_sav...', Array)
#5 /var/www/html/magneto/vendor/magento/framework/Model/AbstractModel.php(678): Magento\Framework\Event\Manager\Proxy->dispatch('sales_order_sav...', Array)
#6 /var/www/html/magneto/vendor/magento/framework/Model/ResourceModel/Db/AbstractDb.php(396): Magento\Framework\Model\AbstractModel->beforeSave()
#7 /var/www/html/magneto/vendor/magento/module-sales/Model/ResourceModel/Order.php(178): Magento\Framework\Model\ResourceModel\Db\AbstractDb->save(Object(Magento\Sales\Model\Order))
#8 /var/www/html/magneto/var/generation/Magento/Sales/Model/ResourceModel/Order/Interceptor.php(24): Magento\Sales\Model\ResourceModel\Order->save(Object(Magento\Sales\Model\Order))
#9 /var/www/html/magneto/vendor/magento/module-sales/Model/OrderRepository.php(161): Magento\Sales\Model\ResourceModel\Order\Interceptor->save(Object(Magento\Sales\Model\Order))
#10 /var/www/html/magneto/vendor/magento/framework/Interception/Interceptor.php(146): Magento\Sales\Model\OrderRepository->save(Object(Magento\Sales\Model\Order))
#11 /var/www/html/magneto/var/generation/Magento/Sales/Model/OrderRepository/Interceptor.php(52): Magento\Sales\Model\OrderRepository\Interceptor->___callPlugins('save', Array, Array)
#12 /var/www/html/magneto/vendor/magento/module-sales/Model/Service/OrderService.php(191): Magento\Sales\Model\OrderRepository\Interceptor->save(Object(Magento\Sales\Model\Order))
#13 /var/www/html/magneto/vendor/magento/module-quote/Model/QuoteManagement.php(488): Magento\Sales\Model\Service\OrderService->place(Object(Magento\Sales\Model\Order))
#14 /var/www/html/magneto/vendor/magento/module-quote/Model/QuoteManagement.php(391): Magento\Quote\Model\QuoteManagement->submitQuote(Object(Magento\Quote\Model\Quote), Array)
#15 /var/www/html/magneto/vendor/magento/module-quote/Model/QuoteManagement.php(349): Magento\Quote\Model\QuoteManagement->submit(Object(Magento\Quote\Model\Quote))
#16 /var/www/html/magneto/vendor/magento/module-checkout/Model/PaymentInformationManagement.php(79): Magento\Quote\Model\QuoteManagement->placeOrder(17)
#17 /var/www/html/magneto/vendor/magento/framework/Interception/Interceptor.php(146): Magento\Checkout\Model\PaymentInformationManagement->savePaymentInformationAndPlaceOrder(17, Object(Magento\Quote\Model\Quote\Payment\Interceptor), Object(Magento\Quote\Model\Quote\Address))
#18 /var/www/html/magneto/var/generation/Magento/Checkout/Model/PaymentInformationManagement/Interceptor.php(26): Magento\Checkout\Model\PaymentInformationManagement\Interceptor->___callPlugins('savePaymentInfo...', Array, Array)
#19 [internal function]: Magento\Checkout\Model\PaymentInformationManagement\Interceptor->savePaymentInformationAndPlaceOrder(17, Object(Magento\Quote\Model\Quote\Payment\Interceptor), Object(Magento\Quote\Model\Quote\Address))
#20 /var/www/html/magneto/vendor/magento/module-webapi/Controller/Rest.php(307): call_user_func_array(Array, Array)
#21 /var/www/html/magneto/vendor/magento/module-webapi/Controller/Rest.php(216): Magento\Webapi\Controller\Rest->processApiRequest()
#22 /var/www/html/magneto/var/generation/Magento/Webapi/Controller/Rest/Interceptor.php(24): Magento\Webapi\Controller\Rest->dispatch(Object(Magento\Framework\App\Request\Http))
#23 /var/www/html/magneto/vendor/magento/framework/App/Http.php(135): Magento\Webapi\Controller\Rest\Interceptor->dispatch(Object(Magento\Framework\App\Request\Http))
#24 /var/www/html/magneto/vendor/magento/framework/App/Bootstrap.php(258): Magento\Framework\App\Http->launch()
#25 /var/www/html/magneto/index.php(39): Magento\Framework\App\Bootstrap->run(Object(Magento\Framework\App\Http))
#26 {main} [] []
```

**Comment by danielneu** [comment]
Hi @lnmn249 

was the module running on 2.1.8 or a lower version?
This could help to pin point the issue.

Which module version are you running, 1.1.5?

Best wishes 
Daniel

**Comment by lnmn249** [comment]
I installed magento 2.1.9 from the get go.  Also installed the module 1.1.4 via the web setup wizard.  I  originally had choices for an earlier version but not 1.1.5.

Dale

**Comment by danielneu** [comment]
Thanks for coming back.

This specific section was updated 5 months ago and not touched since then.
We have no such issues on Magento 2.1.9 or any other version.

I strongly believe another extension is interfering here, which causes this behavior for you.

We can add an additional check to make our implementation more stable, but the consequences will have to be checked first.
For a quick resolution on your side, you can add a check, if the array key "state" exists here, before accessing it:
https://github.com/amzn/amazon-payments-magento-2-plugin/blob/master/src/Payment/Observer/CompleteOrder.php#L58

Something along the lines 

array_key_exists(OrderInterface::STATE, $order->getStoredData());

Best wishes 
Daniel

**Comment by lnmn249** [comment]
Where exactly would you add this code?  My php coding isn't that stellar.

**Comment by danielneu** [comment]
Hi,

a quick shot would be:

        if ($order->getState() == Order::STATE_COMPLETE && 
		    (
				!array_key_exists(OrderInterface::STATE, $order->getStoredData()) ||
				$order->getStoredData()[OrderInterface::STATE] != Order::STATE_COMPLETE) {
				$this->closeOrderReference($order);
			)
        }

Replacing the equivalent section.

As said, we did not investigate if this is the right approach or what we should do, in case there is no stored data available.,
The worst that could happen here should be a CloseOrderReference call too much (will result in us giving an error response back) or too early (will possibly cause a payment that cannot be used for charges,

I would therefore test this first in our sandbox and a staging system and see what the repsonse is.
If you can give me an identifier to look at (S01-xxxxxx-yyyyy number), I can check if you will get into any troubles with it.

If you are not feeling comfortable with this, you will have to wait until a fix is released.

Best wishes 
Daniel

**Comment by lnmn249** [comment]
So.... To ensure I didn't have any configuration or installation errors I started over and installed magento 2.1.9 again and installed the amazon payments module via command line.   It automatically installed 1.1.5, which is what I wanted.   After all the changes I still receive the same error described above.  No other third party modules were installed.  Amazon payments is the only third party module installed.

Then I added the code above with the array_key_exists and went to recompile and received the following error:

```
[magento@payments magento2]$ php bin/magento setup:di:compile
Compilation was started.
Repositories code generation... 1/7 [====>-----------------------]  14% 1 sec 46.0 MiBPHP Parse error:  syntax error, unexpected ';' in /var/www/html/magento2/vendor/amzn/amazon-payments-magento-2-plugin/src/Payment/Observer/CompleteOrder.php on line 68
```
I played around with the code a bit and could not get it to compile.

I know others have this setup working with 1.1.5 and I followed the installation instructions verbatim on the wiki site.  Is there any where else to look for clues as why this isn't working for me?

Dale
